src/dataset.jl

```diff
@@ -0,0 +1,165 @@
+"""
+    Dataset{T}
+
+A `Dataset` describes an associative collection of component `KeyedArray`s with constraints
+on their shared dimensions.
+
+# Fields
+- `constraints::OrderedSet{Pattern}` - Constraint [`Pattern`](@ref)s on shared dimensions.
+- `data::LittleDict{Tuple{Vararg{Symbol}}, T}` - Flattened key paths as tuples of symbols
+  to each component array of type `T`.
+"""
+struct Dataset{T<:XArray}
+    # Our constraints are a collection of pseudo path tuples typically with 1 or
+    # more `:_` wildcard components
+    constraints::OrderedSet{Pattern}
+    # Data lookup can be by any type, but typically it'll either be symbol or tuple.
+    data::LittleDict{Tuple{Vararg{Symbol}}, T}
+end
+
+function Dataset(pairs::Pair{<:Tuple}...; constraints=Pattern[])
+    data = LittleDict{Tuple{Vararg{Symbol}}, XArray}(pairs...)
+
+    # If no constraints have been specified then we default to (:__, dimname)
+    constraint_set = if isempty(constraints)
+        OrderedSet{Pattern}(
+            Pattern(:__, d) for d in Iterators.flatten(dimnames.(values(data)))
+        )
+    else
+        OrderedSet{Pattern}(constraints)
+    end
+
+    result = Dataset(constraint_set, data)
+    validate(result)
+    return result
+end
+
+# Taking pairs is the most general constructor as it doesn't make assumptions about the
+# data key type.
+function Dataset(pairs::Pair{Symbol}...; constraints=Pattern[])
+    data = (
+        Tuple(Symbol.(split(string(k), string(DEFAULT_FLATTEN_DELIM)))) => v
+        for (k, v) in pairs
+    )
+
+    return Dataset(data...; constraints=constraints)
+end
+
+# Utility kwargs constructor.
+Dataset(; constraints=Pattern[], kwargs...) = Dataset(kwargs...; constraints=constraints)
+
+function Base.show(io::IO, ds::Dataset{T}) where T
+    n = length(ds.data)
+    print(io, "Dataset{$T} with $n entries:")
+    for c in ds.constraints
+        print(io, "\n  ", c)
+    end
+    for (k, v) in ds.data
+        printstyled(io, "\n  ", k, " => "; color=:cyan)
+        printstyled(io, replace(sprint(Base.summary, v), "\n" => "\n    "); color=:cyan)
+        print(io, "\n    ", replace(sprint(Base.print_array, v), "\n" => "\n    "))
+    end
+end
+
+#################
+# Dict iterators
+#################
+Base.keys(ds::Dataset) = keys(ds.data)
```

Comment:
Yeah, since both are extending existing functions I'm not sure what the best solution is. I suppose I could just drop `Base.keys` and always do `keys(ds.data)` internally? I'm not sure how often folks will want to use this externally.

Reply:
Alright, we weren't using them that heavily anyways, so I've dropped them until we have a need.